### PR TITLE
Decrease Internal Actor Idle Timeout

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

Yesterday, while taking heap dumps for an App which had run 500 workflow instances in parallel (each with 4 chained activities)  , I had 1 observation that whatever memory was getting allocated to reminders or to workflow objects seemed to get cleared after 1- 1.5 hours.
Was discussing the memory issue with @ItalyPaleAle  and it came out in discussion that it might be due to the Actors idle timeout to be 1 hour.
I ran on local by reducing internal actor's idle timeout to be just 1 minute. And, I could not see reminders or most of the workflow objects in inuse_space after 2-3 minutes of workflows run.

Before: 

image

After:

image

## Issue reference

Please reference the issue this PR is linked to: #6503 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/actors/internal_actor.go [hunk 1]
```diff
@@ -30,6 +31,7 @@ import (
 )
 
 const InternalActorTypePrefix = "dapr.internal."
+const InternalActorIdleTimeout = time.Minute * 1
```
Comment (ItalyPaleAle) [inline pkg/actors/internal_actor.go:34]: With #6538 we will have more internal actors, so hardcoding a timeout may not work for all cases.

WHat about modifying `RegisterInternalActor` so the internal actor can define its own idle timeout?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/actors/internal_actor.go (hunk 1 above)
Comment (DeepanshuA) [reply pkg/actors/internal_actor.go]: Modified accordingly, had to store idle timeout against actor type, so have used `hostedActors` struct for that

File: pkg/actors/actors.go [hunk 2]
```diff
@@ -106,6 +106,9 @@ type Actors interface {
 	CreateTimer(ctx context.Context, req *CreateTimerRequest) error
 	// DeleteTimer deletes an actor timer.
 	DeleteTimer(ctx context.Context, req *DeleteTimerRequest) error
+	IsActorHosted(ctx context.Context, req *ActorHostedRequest) bool
+	GetActiveActorsCount(ctx context.Context) []*runtimev1pb.ActiveActorsCount
+	RegisterInternalActor(ctx context.Context, actorType string, actor InternalActor, actorIdleTimeout time.Duration) error
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:111]: Remove - they are now in the ActorRuntime interface (see #6633 )

```suggestion
```

File: pkg/actors/actors.go [hunk 3]
```diff
@@ -371,6 +374,7 @@ func (a *actorsRuntime) deactivationTicker(configuration Config, deactivateFn de
 				}
 
 				durationPassed := t.Sub(actorInstance.lastUsedTime)
+				log.Infof("Actor type=%s, idleTimeout:%d", actorInstance.actorType, configuration.GetIdleTimeoutForType(actorInstance.actorType))
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go]: Leftover debug log?

File: pkg/actors/config.go [hunk 4]
```diff
@@ -118,6 +118,10 @@ func (c *Config) GetIdleTimeoutForType(actorType string) time.Duration {
 	if val, ok := c.EntityConfigs[actorType]; ok {
 		return val.ActorIdleTimeout
 	}
+	actorIdleTimeout := c.HostedActorTypes.GetActorIdleTimeout(actorType)
+	if actorIdleTimeout != 0 {
```
Comment (ItalyPaleAle) [inline pkg/actors/config.go:122]: Nit:

```suggestion
	if actorIdleTimeout > 0 {
```

File: pkg/actors/internal/config.go [hunk 5]
```diff
@@ -100,3 +103,10 @@ func (ha *hostedActors) ListActorTypes() []string {
 	defer ha.lock.RUnlock()
 	return maps.Keys(ha.actors)
 }
+
+// GetActorIdleTimeout fetches idle timeout stored against an actor type.
+func (ha *hostedActors) GetActorIdleTimeout(actorType string) time.Duration {
+	ha.lock.Lock()
+	defer ha.lock.Unlock()
```
Comment (ItalyPaleAle) [inline pkg/actors/internal/config.go:110]:
```suggestion
	ha.lock.RLock()
	defer ha.lock.RUnlock()
```

Comment (DeepanshuA) [reply pkg/actors/internal/config.go]: done

File: pkg/actors/config.go (hunk 4 above)
Comment (DeepanshuA) [reply pkg/actors/config.go]: done

File: pkg/actors/actors.go (hunk 2 above)
Comment (DeepanshuA) [reply pkg/actors/actors.go:108]: done

Comment (DeepanshuA) [reply pkg/actors/actors.go]: done

**Review by ItalyPaleAle — APPROVED** [review]

**Review by mukundansundar — APPROVED** [review]

**Pull request merged**